### PR TITLE
Fix #666: Disable top-level navigation for data scheme

### LINKED ISSUE #666: [hackerone] #473654 - Address Bar Spoofing
https://hackerone.com/bugs?report_id=473654

### PR DESCRIPTION
**Pull request by jumde**
fix https://github.com/brave/brave-ios/issues/666

Top-level navigation to `data:` urls are not supported on major mobile browsers. On Brave iOS, navigating to data-urls does not work correctly.

Try opening `data:text/html;base64,SGVsbG8gV29ybGQhCg==` in Desktop vs Mobile Browser

This PR removes support for navigating to data URLs.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

### Direct navigation to data:urls
1. Navigate to `data:text/html;base64,SGVsbG8gV29ybGQhCg==` should do a search instead of initiating a navigation to a data URL

### Popups and subresources

1.  Navigate to http://dustiest-limitation.000webhostapp.com/data.html
2. Click on Google.com - data URL should not open
3. Images and CSS rules are loaded correctly.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adaquate test plan exists for QA to validate (if applicable)

### CODE REVIEW

**Comment by jumde** [comment]
cc: @tomlowenthal

**Comment by iccub** [comment]
What more needs to be done in this pr @jumde?

Also I would rename the commit to say something like `disable support for data scheme`, it is a generic change that fixes the security issue as well.

**Comment by jumde** [comment]
@iccub - I'm checking out the test suite to see if any test can be added for this change. 

Updated the commit message, thank you

**Comment by iccub** [comment]
`testSuggestClient` broke it seems

**Comment by iccub** [comment]
Tests work on my local machine, I will rerun Travis

**Review by iccub — APPROVED** [review]
Looks good to me and the commit message is great, thanks.

Approving and waiting for @jhreis for final approval and merge

File: Shared/Extensions/URLExtensions.swift
Comment (jhreis) [inline Shared/Extensions/URLExtensions.swift]: I'm not so sure that we should modify this list, as it is pulled from an external source. What about creating a list of exceptions underneath / above?

Something like:

```
private let permanentURISchemes = // ...
private let ignoredSchemes = ["data"]
private let supportedSchemes = permanentURISchemes.removeAll(where: ignoredSchemes.contains)
```

Then just update `schemeIsValid` to use `supportedSchemes`:

```
    public var schemeIsValid: Bool {
        guard let scheme = scheme else { return false }
        return supportedSchemes.contains(scheme.lowercased())
    }
```

**Review by jhreis — CHANGES_REQUESTED** [review]
Just one small critique. Also we should update the `testschemeIsValid` unit test to include a `"data"` url inside `badurls`.

File: Shared/Extensions/URLExtensions.swift
Comment (jhreis) [inline Shared/Extensions/URLExtensions.swift:107]: Made a mistake previously with the `removeAll`, however the inverse of `filter` is `drop`, and will probably look a bit cleaner:

```suggestion
private let supportedSchemes = permanentURISchemes.drop(where: ignoredSchemes.contains)
```

Comment (jhreis) [reply Shared/Extensions/URLExtensions.swift:107]: This is wrong too. 😄

**Review by jhreis — APPROVED** [review]

**Pull request merged**
